Source/Core/DolphinQt/Config/Graphics/BalloonTip.cpp

```diff
@@ -0,0 +1,299 @@
+// Copyright 2020 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "DolphinQt/Config/Graphics/BalloonTip.h"
+
+#include <QBitmap>
+#include <QGraphicsEffect>
+#include <QGraphicsView>
+#include <QGridLayout>
+#include <QGuiApplication>
+#include <QLabel>
+#include <QPainter>
+#include <QPainterPath>
+#include <QPropertyAnimation>
+#include <QPushButton>
+#include <QStyle>
+
+#if QT_VERSION >= QT_VERSION_CHECK(5, 14, 0)
+#include <QScreen>
+#else
+#include <QApplication>
+#include <QDesktopWidget>
+#endif
+
+namespace
+{
+static BalloonTip* theSolitaryBalloonTip = nullptr;
+}  // namespace
+
+void BalloonTip::showBalloon(const QIcon& icon, const QString& title, const QString& message,
+                             const QPoint& pos, int timeout, bool showArrow, QColor background,
+                             QColor text)
+{
+  hideBalloon();
+  if (message.isEmpty() && title.isEmpty())
+    return;
+
+  theSolitaryBalloonTip = new BalloonTip(icon, title, message, background, text);
+  if (timeout < 0)
+    timeout = 10000;  // 10 s default
+  theSolitaryBalloonTip->balloon(pos, timeout, showArrow);
+
+  QGraphicsOpacityEffect* eff = new QGraphicsOpacityEffect(theSolitaryBalloonTip);
+  theSolitaryBalloonTip->setGraphicsEffect(eff);
+  QPropertyAnimation* a = new QPropertyAnimation(eff, "opacity");
+  a->setDuration(300);
+  a->setStartValue(0);
+  a->setEndValue(1);
+  a->setEasingCurve(QEasingCurve::InExpo);
+  a->start(QPropertyAnimation::DeleteWhenStopped);
+}
+
+void BalloonTip::hideBalloon()
+{
+  if (!theSolitaryBalloonTip)
+    return;
+  theSolitaryBalloonTip->hide();
+  delete theSolitaryBalloonTip;
+  theSolitaryBalloonTip = nullptr;
+}
+
+void BalloonTip::updateBalloonPosition(const QPoint& pos)
+{
+  if (!theSolitaryBalloonTip)
+    return;
+  theSolitaryBalloonTip->hide();
+  theSolitaryBalloonTip->balloon(pos, 0, theSolitaryBalloonTip->showArrow);
+}
+
+bool BalloonTip::isBalloonVisible()
+{
+  return theSolitaryBalloonTip;
+}
+
+BalloonTip::BalloonTip(const QIcon& icon, const QString& title, const QString& message,
+                       QColor background, QColor text)
+    : QWidget(nullptr, Qt::ToolTip), timerId(-1), showArrow(true)
+{
+  setWindowFlags(Qt::FramelessWindowHint);
+  setAttribute(Qt::WA_NoSystemBackground);
+  setAttribute(Qt::WA_TranslucentBackground);
+  setAttribute(Qt::WA_TransparentForMouseEvents);
+  setAttribute(Qt::WA_DeleteOnClose);
+
+#if QT_CONFIG(label)
```

Comment:
This if isn't necessary (ditto elsewhere this is present). A large amount of our code would have broken already if QLabel wasn't present.